runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java

```diff
@@ -912,30 +922,81 @@ public void translateNode(
+                new DoFnOperator.MultiOutputOutputManagerFactory<>(mainTag, outputCoder),
+                windowingStrategy,
+                new HashMap<>(), /* side-input mapping */
+                Collections.emptyList(), /* side inputs */
+                context.getPipelineOptions(),
+                inputKvCoder.getKeyCoder(),
+                keySelector);
+
+        // our operator excepts WindowedValue<KeyedWorkItem> while our input stream
+        // is WindowedValue<SingletonKeyedWorkItem>, which is fine but Java doesn't like it ...
+        @SuppressWarnings("unchecked")
+        SingleOutputStreamOperator<WindowedValue<KV<K, OutputT>>> outDataStream =
+            keyedWorkItemStream
+                .transform(fullName, outputTypeInfo, (OneInputStreamOperator) doFnOperator)
+                .uid(fullName);
+        context.setOutputDataStream(context.getOutput(transform), outDataStream);
+      } else {
+        Tuple2<Map<Integer, PCollectionView<?>>, DataStream<RawUnionValue>> transformSideInputs =
+            transformSideInputs(sideInputs, context);
+
+        TupleTag<KV<K, OutputT>> mainTag = new TupleTag<>("main output");
+        WindowDoFnOperator<K, InputT, OutputT> doFnOperator =
+            new WindowDoFnOperator<>(
+                reduceFn,
+                fullName,
+                (Coder) windowedWorkItemCoder,
+                mainTag,
+                Collections.emptyList(),
+                new DoFnOperator.MultiOutputOutputManagerFactory<>(mainTag, outputCoder),
+                windowingStrategy,
+                transformSideInputs.f0,
+                sideInputs,
+                context.getPipelineOptions(),
+                inputKvCoder.getKeyCoder(),
+                keySelector);
+
+        // we have to manually contruct the two-input transform because we're not
+        // allowed to have only one input keyed, normally.
+
+        TwoInputTransformation<
+                WindowedValue<SingletonKeyedWorkItem<K, InputT>>, RawUnionValue,
+                WindowedValue<KV<K, OutputT>>>
+            rawFlinkTransform =
+                new TwoInputTransformation<>(
+                    keyedWorkItemStream.getTransformation(),
+                    transformSideInputs.f1.broadcast().getTransformation(),
+                    transform.getName(),
+                    (TwoInputStreamOperator) doFnOperator,
+                    outputTypeInfo,
+                    keyedWorkItemStream.getParallelism());
+
+        rawFlinkTransform.setStateKeyType(keyedWorkItemStream.getKeyType());
+        rawFlinkTransform.setStateKeySelectors(keyedWorkItemStream.getKeySelector(), null);
+
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        SingleOutputStreamOperator<WindowedValue<KV<K, OutputT>>> outDataStream =
+            new SingleOutputStreamOperator(
+                keyedWorkItemStream.getExecutionEnvironment(),
+                rawFlinkTransform) {}; // we have to cheat around the ctor being protected
```

Comment:
Wow.

Side note: GitHub is painfully naive about generating diffs for these types of refactor.